### DISCUSSION TITLE
`arglist` and `argv` function behaviours

### DISCUSSION THREAD

**Discussion by mufasachan** [neovim/neovim#29802] [Q&A] [answered]
Hello,

I am trying to script how to handle the output of the `argv` function. In my case `argv` is a single element `.`. 
I use the command `argadd` to add my *real path* transformed argument. But the result is the relative path. So I think `argadd`, `arglist`, perform a simplification step. I look into source to see if `arglist` or `argv` performs specific steps. I failed to understand if there is a transformation or not.

`netrw` perform this gracefully, I looked into its source and I could not find what they do to change the `arglist` correctly. Namely, `netrw`'s `arglist` has real paths when the argument is the `.` directory. *Note*: when the argument is not `.`, `netrw` gives relative path as output of `argv` function. `netrw` is 100% vim script, so I guess I could manage the `arglist` with lua likewise.

So my question is how to set the `arglist`, or anything else the `v:argv` variable for example, to control the output of `argv` function, or `arglist` / `args` contents, please?

Here my `user.lua`
```lua
-- netrw-noload
vim.g.loaded_netrw = 1
vim.g.loaded_netrwPlugin = 1

vim.print("Argv: ", vim.fn.argv(), "----------------------")

local arg_first_real = vim.uv.fs_realpath(vim.fn.argv()[1])
vim.print("Argv (real): ", arg_first_real, "----------------------")

vim.cmd.argdel()
vim.print("Argv: ", vim.fn.argv(), "----------------------")

vim.cmd.argadd(arg_first_real)
vim.print("Argv: ", vim.fn.argv(), "----------------------")
```
The outputs is:
```


Argv: 
{ "debug_argadd.lua" }
----------------------
Argv (real): 
<root_dir>/debug_argadd.lua
----------------------
Argv: 
{}
----------------------
Argv: 
{ "debug_argadd.lua" }
----------------------
```

The nvim command is `nvim -u user.lua .`

```
NVIM v0.10.0
Build type: Release
LuaJIT 2.1.1713484068
```

**Comment by mufasachan** [comment]
I tested with a [recent build](https://github.com/vim/vim/releases/tag/v9.1.0605) of vim and the following command: `VIMRUNTIME=./runtime ./src/vim .` in the repo dir after the `make` command. Inside, I use the command `vim9cmd echo argv()` and I get `.`. The `netrw`'s banner indicates v173, same as the one of neovim.
Plus I diffed, I think the differences are the updates since May 13 2024. And I tested on older vim version, the `netrw` behaviour is not the same. 

Thus, my question is specific to neovim, not vim. I cannot reproduce the behaviour with vim and it might depends of the neovim's implementation, not even `netrw`.

**Comment by zeertzjq** [comment, marked as answer]
When there is a buffer for the file, `argv()` uses the file name of the buffer. When the file is not edited in any buffer, `argv()` uses the file name in the arglist.

**Reply by mufasachan** [reply]
Thanks you very much.
